p2p/src/error.rs

```diff
@@ -14,124 +14,133 @@
+pub enum PublishError {
+    #[error("Message has already been published")]
+    Duplicate,
+    #[error("Failed to sign message")]
+    SigningFailed,
+    #[error("Not enough peers in topic")]
+    InsufficientPeers,
+    #[error("Message is too large")]
+    MessageTooLarge,
+    #[error("Failed to compress the message")]
+    TransformFailed,
+}
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum DialError {
+    #[error("Peer is banned")]
+    Banned,
+    #[error("Limit for outgoing connections reached")]
+    ConnectionLimit,
+    #[error("Tried to dial local node")]
+    LocalPeerId,
+    #[error("Peer doesn't have any known addresses")]
+    NoAddresses,
+    #[error("Peer state not correct for dialing")]
+    DialPeerConditionFalse,
+    #[error("Connection has been aborted")]
+    Aborted,
+    #[error("Invalid PeerId")]
+    InvalidPeerId,
+    #[error("PeerId doesn't match the PeerId of endpoint")]
+    WrongPeerId,
+    #[error("I/O error: `{0:?}`")]
+    IoError(std::io::ErrorKind),
+    #[error("Failed to negotiate transport protocol")]
+    Transport,
+}
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum ConnectionError {
+    #[error("Timeout")]
+    Timeout,
+    #[error("Timer failed")]
+    Timer,
+    #[error("Failed to upgrade protocol")]
+    Upgrade,
 }
 
 #[derive(Error, Debug, PartialEq, Eq)]
 pub enum P2pError {
-    #[error("SocketError: `{0:?}`")]
-    SocketError(std::io::ErrorKind),
-    #[error("PeerDisconnected")]
-    PeerDisconnected,
-    #[error("DecodeFailure: `{0:?}`")]
-    DecodeFailure(String),
-    #[error("ProtocolError: `{0:?}`")]
+    #[error("Protocol violation: `{0:?}`")]
     ProtocolError(ProtocolError),
-    #[error("TimeError: `{0:?}`")]
-    TimeError(String),
-    #[error("Libp2pError: `{0:?}`")]
-    Libp2pError(Libp2pError),
-    #[error("Unknown: `{0:?}`")]
-    Unknown(String),
-    #[error("ChannelClosed")]
+    #[error("Failed to publish message: `{0:?}`")]
+    PublishError(PublishError),
+    #[error("Failed to upgrade connection: `{0:?}`")]
+    ConnectionError(ConnectionError),
+    #[error("Failed to dial peer: `{0:?}`")]
+    DialError(DialError),
+    #[error("Connection to other task lost")]
     ChannelClosed,
-    #[error("NoPeers")]
-    NoPeers,
-    #[error("PeerDoesntExist")]
-    PeerDoesntExist,
-    #[error("InvalidAddress")]
-    InvalidAddress,
-    #[error("InvalidData")]
-    InvalidData,
-    #[error("PeerExists")]
-    PeerExists,
+    #[error("Peer-related error: `{0:?}`")]
+    PeerError(PeerError),
+    #[error("Invalid data: `{0}`")]
+    InvalidData(&'static str),
     #[error("SubsystemFailure")]
     SubsystemFailure,
     #[error("ConsensusError: `{0:?}`")]
     ChainstateError(chainstate::ChainstateError),
     #[error("DatabaseFailure")]
     DatabaseFailure,
-    #[error("InvalidPeerId")]
-    InvalidPeerId,
+    #[error("Failed to convert data `{0}`")]
+    ConversionError(&'static str),
+    #[error("Other: `{0:?}`")]
```

Comment:
This is also just an `&str`. No need for `:?`

Reply:
There are many others. Please try to avoid `:?` as it makes reading logs much harder.